File: conxa-builder/python/services/auth_service.py
"""Clerk OAuth (PKCE) login + token storage for Build Studio.

Flow:
1. Spin up a localhost HTTP server on an OS-assigned port.
2. Open the system browser to Clerk's authorize URL with a PKCE challenge.
3. Catch the redirect, exchange the code for access + refresh tokens.
4. Store tokens in the OS credential manager via ``keyring`` (never plaintext).

Tokens are refreshed transparently when within 60s of expiry. ``get_token``
is what the LLM proxy client calls before each request.
"""

from __future__ import annotations

import base64
import hashlib
import json
import os
import secrets
import threading
import time
import urllib.parse
import urllib.request
import webbrowser
from http.server import BaseHTTPRequestHandler, HTTPServer
from typing import Any

_KEYRING_SERVICE = "conxa-studio"
_TOKEN_KEY = "session"
_REFRESH_LEEWAY_S = 60

_PAGE_CSS = """
*, *::before, *::after { box-sizing: border-box; margin: 0; padding: 0; }
html, body {
  height: 100%;
  background: #0a0c0f;
  font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif;
  color: #e4e4e7;
  display: flex;
  align-items: center;
  justify-content: center;
}
.wrap {
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  gap: 1rem;
  padding: 2rem 1.5rem;
  max-width: 380px;
}
.logo {
  display: flex;
  align-items: center;
  gap: 0.45rem;
  font-size: 1rem;
  font-weight: 600;
  letter-spacing: 0.04em;
  color: #71717a;
  margin-bottom: 1.5rem;
}
.logo-dot {
  width: 9px;
  height: 9px;
  border-radius: 50%;
  background: #3b82f6;
  flex-shrink: 0;
}
.status-icon {
  display: flex;
  align-items: center;
  justify-content: center;
  width: 60px;
  height: 60px;
  border-radius: 50%;
  margin-bottom: 0.75rem;
}
h1 {
  font-size: 1.85rem;
  font-weight: 600;
  color: #f4f4f5;
  letter-spacing: -0.03em;
  line-height: 1.15;
}
.sub {
  font-size: 0.9rem;
  color: #71717a;
  line-height: 1.55;
  margin-top: 0.1rem;
}
.open-btn {
  display: inline-flex;
  align-items: center;
  gap: 0.45rem;
  margin-top: 1rem;
  padding: 0.55rem 1.1rem;
  border-radius: 8px;
  border: 1px solid rgba(255,255,255,0.1);
  background: rgba(255,255,255,0.04);
  color: #a1a1aa;
  font-size: 0.82rem;
  text-decoration: none;
  cursor: pointer;
  transition: background 0.15s, color 0.15s;
}
.open-btn:hover { background: rgba(255,255,255,0.08); color: #e4e4e7; }
"""

_SUCCESS_ICON_SVG = (
    '<svg width="32" height="32" viewBox="0 0 32 32" fill="none">'
    '<circle cx="16" cy="16" r="16" fill="rgba(74,222,128,0.12)"/>'
    '<path d="M9 17L14 22L23 11" stroke="#4ade80" stroke-width="2.5"'
    ' stroke-linecap="round" stroke-linejoin="round"/>'
    "</svg>"
)

_ERROR_ICON_SVG = (
    '<svg width="32" height="32" viewBox="0 0 32 32" fill="none">'
    '<circle cx="16" cy="16" r="16" fill="rgba(248,113,113,0.12)"/>'
    '<path d="M10 10L22 22M22 10L10 22" stroke="#f87171" stroke-width="2.5"'
    ' stroke-linecap="round"/>'
    "</svg>"
)

_OPEN_STUDIO_BTN = (
    '<a href="conxa-studio://open" class="open-btn">'
    '<svg width="14" height="14" viewBox="0 0 14 14" fill="none">'
    '<rect x="1" y="2" width="12" height="9" rx="1.5" stroke="#a1a1aa" stroke-width="1.25"/>'
    '<path d="M5 13h4M7 11v2" stroke="#a1a1aa" stroke-width="1.25" stroke-linecap="round"/>'
    "</svg>"
    "Open Conxa Build Studio"
    "</a>"
)

_DEEP_LINK_SCRIPT = (
    "<script>"
    "setTimeout(function(){window.location.href='conxa-studio://open';},250);"
    "</script>"
)


def _build_callback_html(success: bool, message: str) -> bytes:
    icon = _SUCCESS_ICON_SVG if success else _ERROR_ICON_SVG
    title = "Sign in complete" if success else "Authentication failed"
    subtitle = "You can now close this window." if success else message
    extra = _OPEN_STUDIO_BTN + _DEEP_LINK_SCRIPT if success else ""
    html = (
        "<!DOCTYPE html><html lang='en'><head>"
        "<meta charset='utf-8'/>"
        "<meta name='viewport' content='width=device-width,initial-scale=1'/>"
        f"<title>Conxa — {title}</title>"
        f"<style>{_PAGE_CSS}</style>"
        "</head><body><div class='wrap'>"
        "<div class='logo'><span class='logo-dot'></span>Conxa</div>"
        f"<div class='status-icon'>{icon}</div>"
        f"<h1>{title}</h1>"
        f"<p class='sub'>{subtitle}</p>"
        f"{extra}"
        "</div></body></html>"
    )
    return html.encode()


def _pkce_pair() -> tuple[str, str]:
    verifier = base64.urlsafe_b64encode(secrets.token_bytes(48)).rstrip(b"=").decode()
    challenge = base64.urlsafe_b64encode(
        hashlib.sha256(verifier.encode()).digest()
    ).rstrip(b"=").decode()
    return verifier, challenge


class AuthService:
    def __init__(
        self,
        clerk_domain: str,
        client_id: str,
        *,
        client_secret: str = "",
        cloud_api: str = "",
    ) -> None:
        self._clerk_domain = clerk_domain.rstrip("/")
        self._client_id = client_id
        self._client_secret = client_secret
        self._cloud_api = cloud_api.rstrip("/")
        self._lock = threading.RLock()

    # -- storage -------------------------------------------------------------

    def _keyring(self):
        import keyring  # imported lazily; only present on the desktop build

        return keyring

    def _load(self) -> dict[str, Any] | None:
        raw = self._keyring().get_password(_KEYRING_SERVICE, _TOKEN_KEY)
        if not raw:
            return None
        try:
            return json.loads(raw)
        except json.JSONDecodeError:
            return None

    def _save(self, tokens: dict[str, Any]) -> None:
        self._keyring().set_password(_KEYRING_SERVICE, _TOKEN_KEY, json.dumps(tokens))

    def logout(self) -> None:
        try:
            self._keyring().delete_password(_KEYRING_SERVICE, _TOKEN_KEY)
        except Exception:
            pass

    # -- login ---------------------------------------------------------------

    def login(self, *, on_event=None) -> dict[str, Any]:
        """Run the interactive PKCE login. Returns ``{org_id, user_id, ...}``."""
        if not self._clerk_domain or not self._client_id:
            raise RuntimeError("auth_not_configured")
        verifier, challenge = _pkce_pair()
        state = secrets.token_urlsafe(16)
        result: dict[str, Any] = {}
        done = threading.Event()

        service = self

        class Handler(BaseHTTPRequestHandler):
            def log_message(self, *_a):  # silence default logging
                pass

            def do_GET(self):
                parsed = urllib.parse.urlparse(self.path)
                # Ignore everything except the actual OAuth callback path.
                if parsed.path != "/cb":
                    self.send_response(204)
                    self.end_headers()
                    return
                params = urllib.parse.parse_qs(parsed.query)
                code = (params.get("code") or [""])[0]
                got_state = (params.get("state") or [""])[0]
                error = (params.get("error") or [""])[0]
                error_desc = (params.get("error_description") or [""])[0]
                self.send_response(200)
                self.send_header("Content-Type", "text/html; charset=utf-8")
                self.end_headers()
                if code and got_state == state:
                    self.wfile.write(_build_callback_html(True, "Signed in to Conxa"))
                    result["code"] = code
                elif error:
                    msg = f"{error}: {error_desc}" if error_desc else error
                    result["error"] = msg
                    self.wfile.write(_build_callback_html(False, msg))
                else:
                    result["error"] = "state_mismatch"
                    self.wfile.write(_build_callback_html(False, "Login failed: state mismatch."))
                done.set()

        # Use a fixed port range so the redirect_uri can be pre-registered in Clerk.
        # Random ports can never be pre-registered; Clerk requires an exact URI match.
        _BASE_PORT = int(os.environ.get("CONXA_AUTH_PORT", "52741"))
        server = None
        for _p in range(_BASE_PORT, _BASE_PORT + 10):
            try:
                server = HTTPServer(("127.0.0.1", _p), Handler)
                break
            except OSError:
                continue
        if server is None:
            raise RuntimeError("auth_port_unavailable")
        port = server.server_address[1]
        redirect_uri = f"http://127.0.0.1:{port}/cb"

        authorize = (
            f"{self._clerk_domain}/oauth/authorize?"
            + urllib.parse.urlencode(
                {
                    "response_type": "code",
                    "client_id": self._client_id,
                    "redirect_uri": redirect_uri,
                    "scope": "profile email offline_access user:org:read",
                    "state": state,
                    "code_challenge": challenge,
                    "code_challenge_method": "S256",
                }
            )
        )
        if on_event:
            on_event({"phase": "auth", "step": "browser_open"})
        webbrowser.open(authorize)

        t = threading.Thread(target=server.serve_forever, daemon=True)
        t.start()
        done.wait(timeout=300)
        server.shutdown()

        if not result.get("code"):
            raise RuntimeError(result.get("error") or "login_timeout_or_cancelled")

        tokens = self._exchange_code(result["code"], verifier, redirect_uri)
        tokens["userinfo"] = self._fetch_userinfo_with_retry(tokens["access_token"])
        self._save(tokens)
        return self._claims(tokens)

    def _exchange_code(self, code: str, verifier: str, redirect_uri: str) -> dict[str, Any]:
        params: dict[str, str] = {
            "grant_type": "authorization_code",
            "code": code,
            "client_id": self._client_id,
            "redirect_uri": redirect_uri,
            "code_verifier": verifier,
        }
        if self._client_secret:
            params["client_secret"] = self._client_secret
        body = urllib.parse.urlencode(params).encode()
        return self._token_request(body)

    def _refresh(self, refresh_token: str) -> dict[str, Any]:
        body = urllib.parse.urlencode(
            {
                "grant_type": "refresh_token",
                "refresh_token": refresh_token,
                "client_id": self._client_id,
            }
        ).encode()
        return self._token_request(body)

    def _token_request(self, body: bytes) -> dict[str, Any]:
        url = f"{self._clerk_domain}/oauth/token"
        req = urllib.request.Request(url, data=body, method="POST")
        req.add_header("Content-Type", "application/x-www-form-urlencoded")
        # Cloudflare sits in front of clerk.conxa.in and blocks requests with
        # Python's default user-agent. Mimic a real browser so CF passes it through.
        req.add_header(
            "User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/124.0.0.0 Safari/537.36",
        )
        req.add_header("Accept", "application/json, */*")
        req.add_header("Accept-Language", "en-US,en;q=0.9")
        req.add_header("Origin", self._clerk_domain)
        req.add_header("Referer", f"{self._clerk_domain}/")
        try:
            with urllib.request.urlopen(req, timeout=30) as resp:
                data = json.loads(resp.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            raw = b""
            try:
                raw = exc.read()
            except Exception:
                pass
            body_text = raw.decode("utf-8", errors="replace")
            try:
                err_json = json.loads(body_text)
                clerk_error = err_json.get("error", "")
                clerk_desc = err_json.get("error_description", body_text[:300])
                raise RuntimeError(f"clerk_token_error: {clerk_error}: {clerk_desc}") from exc
            except (ValueError, AttributeError):
                snippet = body_text[:300] or "(empty body)"
                raise RuntimeError(f"clerk_token_http_{exc.code}: {snippet}") from exc
        now = time.time()
        return {
            "access_token": data["access_token"],
            "refresh_token": data.get("refresh_token", ""),
            "exp": now + float(data.get("expires_in", 3600)),
        }

    # -- token access --------------------------------------------------------

    def get_token(self) -> str:
        """Return a valid access token, refreshing if near expiry. Raises if logged out."""
        with self._lock:
            tokens = self._load()
            if not tokens:
                raise RuntimeError("not_authenticated")
            if time.time() >= float(tokens.get("exp", 0)) - _REFRESH_LEEWAY_S:
                refresh = tokens.get("refresh_token")
                if not refresh:
                    raise RuntimeError("session_expired")
                tokens = self._refresh(refresh)
                self._save(tokens)
            return tokens["access_token"]

    def _fetch_userinfo_with_retry(self, access_token: str) -> dict[str, Any]:
        """Fetch userinfo, retrying once after a short delay.

        The freshly-issued access token sometimes isn't propagated through
        Clerk's backend by the time we immediately call /oauth/userinfo.
        A 1-second pause + one retry handles that race without user impact.
        """
        _KEEP = ("sub", "email", "name", "full_name", "org_id")
        last_exc: Exception | None = None
        for attempt in range(2):
            try:
                raw = self._fetch_userinfo(access_token)
                # Only keep the fields we need — Windows Credential Manager
                # has a 2500-byte limit; the full userinfo response exceeds it.
                return {k: raw.get(k) for k in _KEEP}
            except Exception as exc:
                last_exc = exc
                if attempt == 0:
                    time.sleep(1)
        import sys
        print(f"[auth] userinfo fetch failed after retry: {last_exc}", file=sys.stderr)
        return {}

    def _fetch_userinfo(self, access_token: str) -> dict[str, Any]:
        url = f"{self._clerk_domain}/oauth/userinfo"
        req = urllib.request.Request(url, method="GET")
        req.add_header("Authorization", f"Bearer {access_token}")
        req.add_header(
            "User-Agent",
            "Mozilla/5.0 (Windows NT 10.0; Win64; x64) "
            "AppleWebKit/537.36 (KHTML, like Gecko) "
            "Chrome/124.0.0.0 Safari/537.36",
        )
        req.add_header("Accept", "application/json")
        with urllib.request.urlopen(req, timeout=15) as resp:
            return json.loads(resp.read().decode("utf-8"))

    def _claims(self, tokens: dict[str, Any]) -> dict[str, Any]:
        """Extract identity from cached userinfo or fall back to JWT decode."""
        userinfo = tokens.get("userinfo") or {}
        if userinfo:
            return {
                "org_id": userinfo.get("org_id"),
                "user_id": userinfo.get("sub"),
                "name": userinfo.get("name") or userinfo.get("full_name"),
                "email": userinfo.get("email"),
            }
        # Fallback: try to decode as JWT (Clerk may return JWTs on some plans)
        try:
            payload_b64 = tokens["access_token"].split(".")[1]
            payload_b64 += "=" * (-len(payload_b64) % 4)
            claims = json.loads(base64.urlsafe_b64decode(payload_b64))
        except (KeyError, IndexError, ValueError):
            return {}
        return {
            "org_id": claims.get("org_id") or claims.get("orgid"),
            "user_id": claims.get("sub"),
            "name": claims.get("name") or claims.get("full_name"),
            "email": claims.get("email"),
        }

    def current_identity(self) -> dict[str, Any] | None:
        tokens = self._load()
        return self._claims(tokens) if tokens else None

File: conxa-execute/app/electron/auth_service.js
"use strict";
/**
 * Clerk OAuth (PKCE) login for Conxa Execute — uses conxa-cloud's OWN Clerk
 * application (clerk.conxa.in), a separate OAuth client from Build Studio's
 * (its own redirect-URI port range so the two don't collide), NOT a second
 * Clerk instance. Node port of conxa-builder/python/services/auth_service.py's
 * flow: local callback server on a fixed port range (pre-registrable
 * redirect_uri), PKCE S256, token exchange + refresh with a 60s leeway.
 * Tokens are stored via safeStorage.encryptString to userData/clerk-session.bin.
 *
 * Requires CONXA_EXECUTE_CLERK_DOMAIN and CONXA_EXECUTE_CLERK_CLIENT_ID to
 * be set to that OAuth client's values — login throws "auth_not_configured"
 * until these are set. Scope includes user:org:read + keeps org_id in the
 * claims: Execute's personal/team context switcher (execute_client.js's
 * getContexts()) needs to know which Clerk organizations this person
 * belongs to, on top of email/name.
 */
const crypto = require("crypto");
const http = require("http");
const fs = require("fs");
const path = require("path");
const { app, shell, safeStorage } = require("electron");

const REFRESH_LEEWAY_S = 60;
const BASE_PORT = Number(process.env.CONXA_EXECUTE_AUTH_PORT || 52841);
const USER_AGENT =
  "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124.0.0.0 Safari/537.36";

function clerkDomain() {
  return (process.env.CONXA_EXECUTE_CLERK_DOMAIN || "").replace(/\/+$/, "");
}

function clientId() {
  return process.env.CONXA_EXECUTE_CLERK_CLIENT_ID || "";
}

function sessionPath() {
  return path.join(app.getPath("userData"), "clerk-session.bin");
}

function loadTokens() {
  const p = sessionPath();
  if (!fs.existsSync(p)) return null;
  try {
    if (!safeStorage.isEncryptionAvailable()) return null;
    return JSON.parse(safeStorage.decryptString(fs.readFileSync(p)));
  } catch {
    return null;
  }
}

function saveTokens(tokens) {
  if (!safeStorage.isEncryptionAvailable()) {
    throw new Error("OS encryption is not available; the session was not saved.");
  }
  fs.mkdirSync(path.dirname(sessionPath()), { recursive: true });
  fs.writeFileSync(sessionPath(), safeStorage.encryptString(JSON.stringify(tokens)));
}

function logout() {
  try {
    fs.unlinkSync(sessionPath());
  } catch {
    // already logged out
  }
}

function base64url(buf) {
  return buf.toString("base64").replace(/\+/g, "-").replace(/\//g, "_").replace(/=+$/, "");
}

function pkcePair() {
  const verifier = base64url(crypto.randomBytes(48));
  const challenge = base64url(crypto.createHash("sha256").update(verifier).digest());
  return { verifier, challenge };
}

function findServer(handler) {
  return new Promise((resolve, reject) => {
    let p = BASE_PORT;
    const tryListen = () => {
      const server = http.createServer(handler);
      server.on("error", (err) => {
        if (err.code === "EADDRINUSE" && p < BASE_PORT + 10) {
          p += 1;
          tryListen();
        } else {
          reject(err);
        }
      });
      server.listen(p, "127.0.0.1", () => resolve({ server, port: p }));
    };
    tryListen();
  });
}

function transientError(message) {
  const err = new Error(message);
  err.transient = true; // network/server trouble, not "this session is invalid"
  return err;
}

async function tokenRequest(body) {
  const domain = clerkDomain();
  let resp;
  try {
    resp = await fetch(`${domain}/oauth/token`, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": USER_AGENT,
        Accept: "application/json, */*",
        Origin: domain,
        Referer: `${domain}/`,
      },
      body,
    });
  } catch (err) {
    // fetch() rejects on a real network failure (offline, DNS, timeout) —
    // that says nothing about whether the session itself is still good.
    throw transientError(`clerk_token_network_error: ${err.message}`);
  }
  const text = await resp.text();
  let data = null;
  try {
    data = JSON.parse(text);
  } catch {
    // non-JSON error body, handled below
  }
  if (!resp.ok) {
    const desc = data ? data.error_description || data.error : text.slice(0, 300);
    if (resp.status >= 500) throw transientError(`clerk_token_error: ${desc}`);
    throw new Error(`clerk_token_error: ${desc}`);
  }
  const now = Date.now() / 1000;
  return {
    access_token: data.access_token,
    refresh_token: data.refresh_token || "",
    exp: now + Number(data.expires_in || 3600),
  };
}

function exchangeCode(code, verifier, redirectUri) {
  const params = new URLSearchParams({
    grant_type: "authorization_code",
    code,
    client_id: clientId(),
    redirect_uri: redirectUri,
    code_verifier: verifier,
  });
  return tokenRequest(params.toString());
}

function refreshTokens(refreshToken) {
  const params = new URLSearchParams({
    grant_type: "refresh_token",
    refresh_token: refreshToken,
    client_id: clientId(),
  });
  return tokenRequest(params.toString());
}

async function fetchUserinfo(accessToken) {
  const resp = await fetch(`${clerkDomain()}/oauth/userinfo`, {
    headers: { Authorization: `Bearer ${accessToken}`, "User-Agent": USER_AGENT, Accept: "application/json" },
  });
  if (!resp.ok) throw new Error(`userinfo_http_${resp.status}`);
  return resp.json();
}

async function fetchUserinfoWithRetry(accessToken) {
  // The freshly-issued access token sometimes isn't propagated through
  // Clerk's backend by the time we immediately call /oauth/userinfo — one
  // retry after a short pause handles that race (mirrors auth_service.py).
  const keep = ["sub", "email", "name", "full_name", "org_id"];
  for (let attempt = 0; attempt < 2; attempt++) {
    try {
      const raw = await fetchUserinfo(accessToken);
      const out = {};
      for (const k of keep) out[k] = raw[k];
      return out;
    } catch {
      if (attempt === 0) await new Promise((r) => setTimeout(r, 1000));
    }
  }
  return {};
}

function claimsFromTokens(tokens) {
  const userinfo = tokens.userinfo || {};
  if (userinfo.sub) {
    return { user_id: userinfo.sub, name: userinfo.name || userinfo.full_name, email: userinfo.email };
  }
  // Fallback: decode as JWT (Clerk may return JWTs on some plans).
  try {
    const payloadB64 = tokens.access_token.split(".")[1];
    const payload = JSON.parse(Buffer.from(payloadB64, "base64").toString("utf8"));
    if (!payload.sub) return null;
    return { user_id: payload.sub, name: payload.name || payload.full_name, email: payload.email };
  } catch {
    return null;
  }
}

async function login() {
  if (!clerkDomain() || !clientId()) {
    // ponytail: dev-only bypass, real Clerk app not wired up yet — remove once CONXA_EXECUTE_CLERK_DOMAIN/CLIENT_ID are set
    if (!require("electron").app.isPackaged) {
      const userinfo = { sub: "dev-user", name: "Dev User", email: "dev@conxa.local" };
      saveTokens({ access_token: "dev-bypass-token", refresh_token: "", exp: Date.now() / 1000 + 86400 * 365, userinfo });
      return claimsFromTokens({ userinfo });
    }
    throw new Error("auth_not_configured");
  }
  const { verifier, challenge } = pkcePair();
  const state = base64url(crypto.randomBytes(16));

  let resolveCode, rejectCode;
  const codePromise = new Promise((res, rej) => {
    resolveCode = res;
    rejectCode = rej;
  });

  const { server, port } = await findServer((req, res) => {
    const url = new URL(req.url, "http://127.0.0.1");
    if (url.pathname !== "/cb") {
      res.writeHead(204);
      res.end();
      return;
    }
    const code = url.searchParams.get("code");
    const gotState = url.searchParams.get("state");
    const error = url.searchParams.get("error");
    res.writeHead(200, { "Content-Type": "text/html; charset=utf-8" });
    if (code && gotState === state) {
      res.end("<html><body><h1>Signed in to CONXA</h1><p>You can close this window.</p></body></html>");
      resolveCode(code);
    } else {
      const msg = error || "state_mismatch";
      res.end(`<html><body><h1>Sign-in failed</h1><p>${msg}</p></body></html>`);
      rejectCode(new Error(msg));
    }
  });

  const redirectUri = `http://127.0.0.1:${port}/cb`;
  const authorizeUrl =
    `${clerkDomain()}/oauth/authorize?` +
    new URLSearchParams({
      response_type: "code",
      client_id: clientId(),
      redirect_uri: redirectUri,
      scope: "profile email offline_access user:org:read",
      state,
      code_challenge: challenge,
      code_challenge_method: "S256",
    }).toString();

  await shell.openExternal(authorizeUrl);

  let code;
  try {
    code = await Promise.race([
      codePromise,
      new Promise((_resolve, reject) => setTimeout(() => reject(new Error("login_timeout")), 300_000)),
    ]);
  } finally {
    server.close();
  }

  const tokens = await exchangeCode(code, verifier, redirectUri);
  tokens.userinfo = await fetchUserinfoWithRetry(tokens.access_token);
  const identity = claimsFromTokens(tokens);
  if (!identity || !identity.user_id) throw new Error("login_identity_missing");
  saveTokens(tokens);
  return identity;
}

async function getToken() {
  let tokens = loadTokens();
  if (!tokens) throw new Error("not_authenticated");
  if (Date.now() / 1000 >= Number(tokens.exp || 0) - REFRESH_LEEWAY_S) {
    if (!tokens.refresh_token) throw new Error("session_expired");
    const oldUserinfo = tokens.userinfo || {};
    tokens = await refreshTokens(tokens.refresh_token);
    tokens.userinfo = Object.keys(oldUserinfo).length ? oldUserinfo : await fetchUserinfoWithRetry(tokens.access_token);
    saveTokens(tokens);
  }
  return tokens.access_token;
}

async function currentIdentity() {
  const cached = loadTokens();
  if (!cached) return null;
  try {
    await getToken();
  } catch (err) {
    if (err.transient) {
      // A network blip or Clerk 5xx during refresh doesn't mean the user
      // signed out — show them as still signed in using the last-known
      // identity rather than bouncing them to the sign-in screen.
      const identity = claimsFromTokens(cached);
      return identity && identity.user_id ? identity : null;
    }
    return null; // refresh token missing/expired/revoked — genuinely signed out
  }
  const tokens = loadTokens();
  const identity = tokens ? claimsFromTokens(tokens) : null;
  return identity && identity.user_id ? identity : null;
}

module.exports = { login, logout, getToken, currentIdentity };
